### DISCUSSION TITLE
What's wrong in this prepipe command?

### DISCUSSION THREAD

**Discussion by aborruso** [johnkerl/miller#539] [Q&A] [answered]
Hi,
[this](https://github.com/johnkerl/miller/files/6546990/piemonte.zip) zip file contains only one CSV file in a subfolder.

I'm trying to read it using `mlr --prepipe 'unzip -qc' --csv --ifs "|" cat piemonte.zip`, but it does not work.

If I run simply `unzip -qc piemonte.zip`  I have the CSV in the stdout.

Thank you

**Comment by aborruso** [comment]
@johnkerl do you know what's wrong?

Thank you very much

**Reply by johnkerl** [reply]
Just now saw this, looking ...

**Comment by johnkerl** [comment, marked as answer]
@aborruso what Miller does with `mlr --prepipe somecommand file.dat` is to `popen` from the process `somecommand < file.dat` -- note the `<`. And it turns out `unzip -qc piemonte.zip` works while `unzip -qc < piemonte.zip` does not.

This makes `unzip -qc` different from most other commands like `gunzip` which are happy with `<`.

In fact this is an important distinction to preserve since `gunzip foo.gz` isn't the same as `gunzip < foo.gz` -- the latter leaves `foo.gz in place; the former removes it and replaces it with the decompressed `foo`.

What we need is a `--prepipe-no-less-than` or something.

**Reply by aborruso** [reply]
Thank you very much. Now I have to find some unzip command that uses `<` and that is able to uncompress a single file that is inside a subfolder.

**Reply by aborruso** [reply]
@johnkerl I have made another test using `bsdtar`. It's able to use `<` but it extracts to current folder not to `stdout`

```
bsdtar -xf- -s'|[^/]*/||' <piemonte.zip
```

I think I must use a programmatically way.

Thank you

**Reply by johnkerl** [reply]
@aborruso this is fixed (for Miller 6 anyway) in https://github.com/johnkerl/miller/pull/554

```
$ mlr --prepipex 'unzip -qc' --icsv --ifs pipe --oxtab tail -n 2 piemonte.zip
id     8
campi1 come stai
lat    38.123456
lon

id     9
campi1 come stai
lat    38.123456
lon    13.123456
```

**Reply by aborruso** [reply]
Wow thank you very much

**Reply by aborruso** [reply]
@johnkerl to test it, do I must build the go version?

**Reply by johnkerl** [reply]
@aborruso , or I can make you a binary if you like.

**Reply by johnkerl** [reply]
@aborruso I thought I had https://github.com/johnkerl/miller/actions set up to retain binaries but I have something to fix there. :(

Also -- if this is blocking you I can backport it to the C implementation (Miller 5.x)

**Reply by aborruso** [reply]
> @aborruso , or I can make you a binary if you like.

I'll try to do it. If I can't do it, I'll write to you.

Thank you

**Reply by johnkerl** [reply]
@aborruso there are Linux and MacOS binaries at https://github.com/johnkerl/miller/actions/runs/903075460#artifacts -- I have something to fix with the Windows version though

**Reply by aborruso** [reply]
> @aborruso there are Linux and MacOS binaries at https://github.com/johnkerl/miller/actions/runs/903075460#artifacts

Great, thank you

**Reply by johnkerl** [reply]
@aborruso all three including Windows! :D 
https://github.com/johnkerl/miller/actions/runs/903119077#artifacts
